allowances/README.md

```diff
@@ -0,0 +1,122 @@
+
+### Considerations
+
+- Transactions can be front-run/ executed by anyone once they are published to the mempool. A way to prevent this would be to introduce a `paymentReceiver`.
+- The payment logic adds additional code and could have been removed for an initial version. 
+
+## Architecture
+
+The contract has been designed to improve the gas usage when executing a transfer and to make it easy to interact with the contract. Because of this configuration changes (e.g. adding or removing allowance and delegates) are not optimized on gas usage.
+
+Allowances are store in a map called `allowances`. The key to this map is a combination of the Safe, the delegate and the token. Each allowance has the following fields:
+- `uint96 amount`
+  - Maximum amount that can be spent
+- `uint96 spent`
+  - Amount that has already been spent
+- `uint16 resetTimeMin`
+  - Time span in **minutes** after which the `spent` amount will automatically reset to `0`. If `resetTimeMin` is `0` the `spent` value will not automaticall reset.
+- `uint32 lastResetMin`
+  - Point in time in **minutes** when the `spent` amount was last reset to `0`. This can be used to specify the point in time (e.g. time of a day) when the reset should happen, as each following reset will be based on last time.
+- `uint16 nonce`
+  - Increasing nonce that is used to protect transfer confirmations (e.g. signatures) against replay attacks.
+
+Some limits have been applied to the different fields to make sure that the overall struct fits into a single **word** of storage:
+- The maximum allowance is `2^96`. Assuming a token with 18 decimals that would allow an allowance from more than 70 billion, which seems sufficient.
+- The maximum time span that can be used for the reset is `2**16` minutes. This corresponds to around 45 days, which is more than one month and should be sufficient for most use cases.
+- The maximum reset time would be the unix epoch of `(2**32) * 60`. This is in the year 10136.
+- The maximum amount of transfer a delegate can do is `2**16` (aka `65536`) which should be sufficient, as it is easily possible to register a new delegate.
+
+To make it easier to allow querying for allowances two more data structures have been added to the contract: `delegates` and `tokens`.
+
+`delegates` stores all the delegates for a specific Safe in a double linked list. For this we also store an entry point to that double linked list in `delegatesStart`. Each delegate is identified by a `uint48` which is the first **6bytes** of the delegate address. Because of this is can theoretically come to collisions. As the index points to a struct containing the `address` of the delgate, the `next` index and the `prev` index, it is possible to verify which address was used to get the index. In case of collisions we recommend to generate a new delegate.
```

Comment:
~Because of this is can theoretically come to collisions.~

This could theoretically cause collisions.